p2p/src/net/mock/transport/channel.rs

```diff
@@ -0,0 +1,177 @@
+// Copyright (c) 2022 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::{collections::BTreeMap, io, sync::Mutex};
+
+use async_trait::async_trait;
+use once_cell::sync::Lazy;
+use tokio::sync::{
+    mpsc::{unbounded_channel, UnboundedReceiver, UnboundedSender},
+    oneshot::{self, Sender},
+};
+
+use crate::{
+    error::DialError,
+    net::mock::{
+        transport::{MockListener, MockStream, MockTransport},
+        types::Message,
+    },
+    P2pError, Result,
+};
+
+type Address = u64;
+type MessageSender = UnboundedSender<Message>;
+type MessageReceiver = UnboundedReceiver<Message>;
+type AcceptResponse = (MessageSender, MessageReceiver);
+
+/// Zero address has special meaning: bind to a free address.
+const ZERO_ADDRESS: Address = 0;
+
+static CONNECTIONS: Lazy<Mutex<BTreeMap<Address, UnboundedSender<Sender<AcceptResponse>>>>> =
+    Lazy::new(|| Mutex::new(BTreeMap::new()));
+
+#[derive(Debug)]
+pub struct ChannelMockTransport {}
+
+#[async_trait]
+impl MockTransport for ChannelMockTransport {
+    type Address = Address;
+    type Listener = ChannelMockListener;
+    type Stream = ChannelMockStream;
+
+    async fn bind(address: Self::Address) -> Result<Self::Listener> {
+        let mut connections = CONNECTIONS.lock().expect("Connections mutex is poisoned");
+
+        let address = if address == ZERO_ADDRESS {
+            connections.iter().next_back().map_or(1, |(&a, _)| a + 1)
+        } else {
+            address
+        };
+
+        let (sender, receiver) = unbounded_channel();
+        assert!(connections.insert(address, sender).is_none());
+
+        Ok(Self::Listener { address, receiver })
+    }
+
+    async fn connect(address: Self::Address) -> Result<Self::Stream> {
+        // A connection can only be established to a known address.
+        debug_assert_ne!(ZERO_ADDRESS, address);
```

Comment:
This is a mock... why debug?

Reply:
I thought that having an assertion here isn't strictly necessary because this function returns an error if zero address is used anyway. But passing such address here indicated a developer error, so `debug_assert` feels appropriate. Fixed.